### PR TITLE
tests: Add unit tests for rpc client endpoints

### PR DESCRIPTION
**Pull request by Vvaradinov**
Add additional unit tests to achieve >90% code coverage and expand the mock clients.

Closes: [1248](https://github.com/evmos/ethermint/issues/1248)
Issues: [1330](https://github.com/evmos/ethermint/issues/1330), [1331](https://github.com/evmos/ethermint/issues/1331), [1332](https://github.com/evmos/ethermint/issues/1332), [1333](https://github.com/evmos/ethermint/issues/1333), [1334](https://github.com/evmos/ethermint/issues/1334), [1335](https://github.com/evmos/ethermint/issues/1335), [1336](https://github.com/evmos/ethermint/issues/1336)

## Description

______

For contributor use:

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [ ] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added  relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

______

For admin use:

- [ ] Added appropriate labels to PR (ex. `WIP`, `R4R`, `docs`, etc)
- [ ] Reviewers assigned
- [ ] Squashed all commits, uses message "Merge pull request #XYZ: [title]" ([coding standards](https://github.com/tendermint/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: rpc/backend/utils.go [hunk 1]
```diff
@@ -47,6 +47,7 @@ func (s sortGasAndReward) Less(i, j int) bool {
 func (b *Backend) getAccountNonce(accAddr common.Address, pending bool, height int64, logger log.Logger) (uint64, error) {
 	queryClient := authtypes.NewQueryClient(b.clientCtx)
 	res, err := queryClient.Account(types.ContextWithHeight(height), &authtypes.QueryAccountRequest{Address: sdk.AccAddress(accAddr.Bytes()).String()})
+	fmt.Println("err account query", err, "res", res)
```
Comment (fedekunze) [inline rpc/backend/utils.go:50]:
```suggestion
```

**Comment by Vvaradinov** [comment]
Coverage currently achieved is about 74%. An additional PR will be made to resolve an issue regarding mocking a `RegisterABCIQueryAccount`  which once fixed will bump the coverage to > 80%. Unfortunately some functions make checks that are not reachable / mockable so coverage on some files is between 70-80%.

File: rpc/backend/backend_suite_test.go [hunk 2]
```diff
@@ -157,3 +169,31 @@ func (suite *BackendTestSuite) generateTestKeyring(clientDir string) (keyring.Ke
...
+	suite.Require().NoError(err)
+
+	from := common.BytesToAddress(priv.PubKey().Address().Bytes())
+	signer := tests.NewSigner(priv)
+
+	queryClient := suite.backend.queryClient.QueryClient.(*mocks.EVMQueryClient)
+	RegisterParamsWithoutHeader(queryClient, 1)
+
+	armor := crypto.EncryptArmorPrivKey(priv, "", "eth_secp256k1")
+	suite.backend.clientCtx.Keyring.ImportPrivKey("test_key", armor, "")
```
Comment (fedekunze) [inline rpc/backend/backend_suite_test.go:184]: Looks like this can be moved to the test suite setup

File: rpc/backend/backend_suite_test.go [hunk 3]
```diff
@@ -157,3 +169,31 @@ func (suite *BackendTestSuite) generateTestKeyring(clientDir string) (keyring.Ke
 	encCfg := encoding.MakeConfig(app.ModuleBasics)
 	return keyring.New(sdk.KeyringServiceName(), keyring.BackendTest, clientDir, buf, encCfg.Codec, []keyring.Option{hd.EthSecp256k1Option()}...)
 }
+
+func (suite *BackendTestSuite) signAndEncodeEthTx(msgEthereumTx *evmtypes.MsgEthereumTx) ([]byte, error) {
+	priv, err := ethsecp256k1.GenerateKey()
+	suite.Require().NoError(err)
+
+	from := common.BytesToAddress(priv.PubKey().Address().Bytes())
```
Comment (fedekunze) [inline rpc/backend/backend_suite_test.go:177]: there is a `tests` util package that can help generate this

File: rpc/backend/backend_suite_test.go [hunk 4]
```diff
@@ -157,3 +169,31 @@ func (suite *BackendTestSuite) generateTestKeyring(clientDir string) (keyring.Ke
...
+	suite.Require().NoError(err)
+
+	tx, err := msgEthereumTx.BuildTx(suite.backend.clientCtx.TxConfig.NewTxBuilder(), "aphoton")
+	suite.Require().NoError(err)
+
+	txEncoder := suite.backend.clientCtx.TxConfig.TxEncoder()
+	txBz, err := txEncoder(tx)
+	suite.Require().NoError(err)
+
+	return txBz, nil
```
Comment (fedekunze) [inline rpc/backend/backend_suite_test.go:198]: this is always returning nil error, we should only return the tx bytes

File: rpc/backend/backend.go [hunk 5]
```diff
@@ -162,22 +159,23 @@ func NewBackend(
...
-		)
-		if err != nil {
-			panic(err)
-		}
-
-		clientCtx = clientCtx.WithKeyring(kr)
-	}
+	// TODO: Figure out why this was in the original code
+	//algos, _ := clientCtx.Keyring.SupportedAlgorithms()
+	//if !algos.Contains(hd.EthSecp256k1) {
```
Comment (fedekunze) [inline rpc/backend/backend.go:164]: fix linter

File: rpc/backend/call_tx_test.go [hunk 6]
```diff
@@ -0,0 +1,484 @@
...
+			nil,
+			common.Hash{},
+			false,
+		},
+		{
+			"pass - Can't set Tx defaults BaseFee disabled",
+			func() {
+				client := suite.backend.clientCtx.Client.(*mocks.Client)
+				queryClient := suite.backend.queryClient.QueryClient.(*mocks.EVMQueryClient)
+				var header metadata.MD
```
Comment (fedekunze) [inline rpc/backend/call_tx_test.go:62]: nit, vars should be defined at the top of the function

File: rpc/backend/call_tx_test.go [hunk 7]
```diff
@@ -0,0 +1,484 @@
...
+		Gas:                  nil,
+		GasPrice:             nil,
+		MaxFeePerGas:         gasPrice,
+		MaxPriorityFeePerGas: gasPrice,
+		Value:                gasPrice,
+		Input:                nil,
+		Data:                 nil,
+		AccessList:           nil,
+	}
+	argsBz, _ := json.Marshal(callArgs)
```
Comment (fedekunze) [inline rpc/backend/call_tx_test.go:370]: nit, check error

File: rpc/backend/call_tx_test.go [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+		callArgs     evmtypes.TransactionArgs
+		expEthTx     *evmtypes.MsgEthereumTxResponse
+		expPass      bool
+	}{
+		{
+			"fail - Invalid request",
+			func() {
+				client := suite.backend.clientCtx.Client.(*mocks.Client)
+				queryClient := suite.backend.queryClient.QueryClient.(*mocks.EVMQueryClient)
+				//var header metadata.MD
```
Comment (fedekunze) [inline rpc/backend/call_tx_test.go:385]:
```suggestion
```

File: rpc/backend/chain_info_test.go [hunk 9]
```diff
@@ -144,3 +150,299 @@ func (suite *BackendTestSuite) TestBaseFee() {
...
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.Run(fmt.Sprintf("case %s", tc.name), func() {
+			suite.SetupTest() // reset test and queries
+			tc.registerMock(tc.validator)
+
+			feeHistory, err := suite.backend.FeeHistory(tc.userBlockCount, tc.latestBlock, []float64{25, 50, 75, 100})
+			suite.T().Log(feeHistory, err)
```
Comment (fedekunze) [inline rpc/backend/chain_info_test.go:439]:
```suggestion
```

File: rpc/backend/node_info_test.go [hunk 10]
```diff
@@ -0,0 +1,359 @@
...
+		expAddr      []common.Address
+		expPass      bool
+	}{
+		{
+			"pass - returns empty address",
+			func() {},
+			[]common.Address{},
+			true,
+		},
+		// TODO: Manual address adding to keyring
```
Comment (fedekunze) [inline rpc/backend/node_info_test.go:110]: Is this for the follow up PR?

File: rpc/backend/node_info_test.go [hunk 11]
```diff
@@ -0,0 +1,359 @@
...
+					"/cosmos.auth.v1beta1.Query/Account",
+					requestMarshal,
+					tmrpcclient.ABCIQueryOptions{Height: int64(1), Prove: false},
+				)
+			},
+			common.Address{},
+			false,
+		},
+		//{
+		//	"pass - set the etherbase for the miner",
```
Comment (fedekunze) [inline rpc/backend/node_info_test.go:274]: ditto

File: rpc/backend/node_info_test.go [hunk 12]
```diff
@@ -0,0 +1,359 @@
...
+
+	for _, tc := range testCases {
+		suite.Run(fmt.Sprintf("case %s", tc.name), func() {
+			suite.SetupTest() // reset test and queries
+			tc.registerMock()
+
+			output := suite.backend.SetEtherbase(tc.etherbase)
+
+			//suite.T().Error(output)
+
```
Comment (fedekunze) [inline rpc/backend/node_info_test.go:307]:
```suggestion
```

File: rpc/backend/sign_tx_test.go [hunk 13]
```diff
@@ -0,0 +1,267 @@
...
+	"google.golang.org/grpc/metadata"
+)
+
+func (suite *BackendTestSuite) TestSendTransaction() {
+	gasPrice := new(hexutil.Big)
+	gas := hexutil.Uint64(1)
+	zeroGas := hexutil.Uint64(0)
+	toAddr := tests.GenerateAddress()
+	priv, _ := ethsecp256k1.GenerateKey()
+	from := common.BytesToAddress(priv.PubKey().Address().Bytes())
```
Comment (fedekunze) [inline rpc/backend/sign_tx_test.go:26]: you can also generate keys using the tests package

File: rpc/backend/sign_tx_test.go [hunk 14]
```diff
@@ -0,0 +1,267 @@
...
+			} else {
+				suite.Require().Error(err)
+			}
+		})
+	}
+}
+
+func (suite *BackendTestSuite) TestSignTypedData() {
+	priv, _ := ethsecp256k1.GenerateKey()
+	from := common.BytesToAddress(priv.PubKey().Address().Bytes())
```
Comment (fedekunze) [inline rpc/backend/sign_tx_test.go:219]: ditto

File: rpc/backend/tx_info_test.go [hunk 15]
```diff
@@ -0,0 +1,599 @@
...
+
+			db := dbm.NewMemDB()
+			suite.backend.indexer = indexer.NewKVIndexer(db, tmlog.NewNopLogger(), suite.backend.clientCtx)
+			err := suite.backend.indexer.IndexBlock(block, responseDeliver)
+			suite.Require().NoError(err)
+
+			rpcTx, err := suite.backend.GetTransactionByHash(common.HexToHash(tc.tx.Hash))
+
+			//suite.T().Error(rpcTx, tc.expRPCTx)
+			//suite.T().Error(err)
```
Comment (fedekunze) [inline rpc/backend/tx_info_test.go:115]:
```suggestion
```

**Review by fedekunze — COMMENTED** [review]
great work! Left comments for some improvements 🚀

**Review by facs95 — APPROVED** [review]

**Pull request merged**